### ISSUE TITLE
TypeVar is not subscriptable

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#8255]
**Describe the bug**

Pyright incorrectly reports `TypeVar "Baz" is not subscriptable` at an annotation like `Baz[T, U]`. Removing the type arguments in another location results in an error `Expected type arguments for generic type alias "Baz"`.

The first issue is the message says Baz is a TypeVar though it's not (the hover text is right). 

The second issue appears to be due to pyright not recognizing recursive types. Writing `from __future__ import annotations` at the top of each file gets rid of the false error.

**Code or Screenshots**

I've created a minimal reproducible example [here](https://github.com/kyle-query/pyright-bug/tree/main). The order in which files are analyzed seems to determine which file, either `foo.py` or `bar.py`, is the source of the false error.

**VS Code extension or command-line**

The issue occurs with pyright 1.1.369 as well as the latest release of pylance.

**Comment by erictraut** [comment]
This error is due to a circular dependency in your code. `Baz` is defined in terms of `BazAtters` and `BazOther` which depends on `Foo` which depends on `FooOther` which depends on `Baz`. For most classes, the types of attributes would not need to be resolved at the time a class type is evaluated, but for `TypedDict` and `dataclass` classes, attribute types must be analyzed earlier immediately because an `__init__` method needs to be synthesized, and class decorators can depend on that `__init__` method.

The error message that pyright emits here is admittedly confusing. I could modify the code to be more aggressive about detecting and reporting such circularities, but it would unnecessarily flag some cases that pyright is otherwise able to handle without an error. So I don't think that's a good tradeoff.

In summary, I don't consider this a bug in pyright. You'll need to find a way to break this circularity in your code if you want this to work with pyright.

**Comment by kyle-query** [comment]
Thanks Eric. Does it make sense that the errors aren't reported when I have `from __future__ import annotations`? I understood PEP 563 as a way to avoid evaluating type annotations at runtime, so I'm surprised it "fixes" the type error.

I also notice the error doesn't occur if I put all the definitions in the same file, either writing `a: "Baz[Any, U]"` or using `from __future__ import annotations`. I pushed another file into that repo called example/zzz.py that demonstrates this.

Are either of these one of the cases you referred to which pyright can handle without reporting an error? Writing the annotation in another file `x: Bar[int, str]` seems to work as expected.
